### DISCUSSION TITLE
Does scroll restoration work

### DISCUSSION THREAD

**Discussion by L-Blondy** [TanStack/virtual#677] [Q&A] [answered]
Hi there, 

I'm considering using tanstack virtual at work and scroll restoration is a requirement. Does it work properly when navigating back / forward?

I could not find anything about it in the docs.

Thanks!

**Comment by KonstantinKai** [comment, marked as answer]
Yep, just use the following options while creating an instance of `virtualized`
```ts
initialOffset: number;
initialMeasurementsCache: VirtualItem[],
```

Here is my snippet and it works fine for me:

```ts
let _kSavedOffset = 0;
let _KMeasurementsCache = [] as VirtualItem[];

const Element: React.FC = () => {
  const virtualized = useVIrtualizer({
    // options,
    initialOffset: _kSavedOffset,
    initialMeasurementsCache: _kMeasurementsCache,
    onChange: (virtualizer) => {
      if (!virtualizer.isScrolling) {
        _kMeasurementsCache = virtualizer.measurementsCache;
        _kSavedOffset = virtualizer.scrollOffset;
      }
    },
  });
  return <div>{/*your list goes here*/}</div>
}
```

**Reply by younessbennaj** [reply]
Thank you 🙏

**Reply by Frans06** [reply]
for those using dynamic elements I recomment always set the element height if horizontal or element width if vertical. I use the following code.
https://github.com/TanStack/virtual/issues/651#issuecomment-2439090341
